File: exercises/domain-5/block-5.3/starter.ts
interface TaskContext {
  description: string
  filesModified: string[]
  includesDbMigration: boolean
  isDeployment: boolean
  linesChanged: number
}

interface ReviewGate {
  name: string
  trigger: (ctx: TaskContext) => boolean
  message: (ctx: TaskContext) => string
}

class ReviewGatedAgent {
  private gates: ReviewGate[] = []
  gateLog: Array<{ gate: string; triggered: boolean; taskDescription: string }> = []

  addGate(gate: ReviewGate): this {
    this.gates.push(gate)
    return this
  }

  // TODO: Evaluate all gates against the context
  // If any gate triggers, show review and wait for approval
  // Return true if approved/no gates triggered, false if rejected
  async evaluateGates(ctx: TaskContext, getApproval: (message: string) => Promise<boolean>): Promise<boolean> {
    return true
  }

  // TODO: Run a task with gate evaluation
  async runTask(ctx: TaskContext, task: () => Promise<void>, getApproval: (message: string) => Promise<boolean>): Promise<void> {
    console.log(`Running task: ${ctx.description}`)
  }
}

// Demo
const agent = new ReviewGatedAgent()

// Add gates
agent
  .addGate({
    name: "many-files",
    trigger: (ctx) => ctx.filesModified.length > 5,
    message: (ctx) => `⚠️ This task modifies ${ctx.filesModified.length} files. Please review the file list.`,
  })
  .addGate({
    name: "db-migration",
    trigger: (ctx) => ctx.includesDbMigration,
    message: () => "⚠️ Database schema change detected. Migrations cannot be easily reversed.",
  })
  .addGate({
    name: "deployment",
    trigger: (ctx) => ctx.isDeployment,
    message: () => "🚀 Deployment gate: This will push changes to production.",
  })

async function simulateApproval(message: string): Promise<boolean> {
  console.log(`\n[Review Required]\n${message}\n[Auto-approving for demo: YES]`)
  return true
}

async function main() {
  // Task 1: Small change (no gates)
  await agent.runTask(
    { description: "Fix typo in README", filesModified: ["README.md"], includesDbMigration: false, isDeployment: false, linesChanged: 1 },
    async () => console.log("  ✓ Typo fixed"),
    simulateApproval
  )

  // Task 2: Large refactor (triggers many-files gate)
  await agent.runTask(
    { description: "Refactor auth module", filesModified: Array.from({length: 8}, (_, i) => `auth/file${i}.ts`), includesDbMigration: false, isDeployment: false, linesChanged: 340 },
    async () => console.log("  ✓ Auth refactored"),
    simulateApproval
  )

  console.log("\nGate log:", agent.gateLog)
}

main().catch(console.error)

File: exercises/domain-5/block-5.3/starter.py
import asyncio
from dataclasses import dataclass, field
from typing import Callable, Awaitable


@dataclass
class TaskContext:
    description: str
    files_modified: list[str]
    includes_db_migration: bool
    is_deployment: bool
    lines_changed: int


@dataclass
class ReviewGate:
    name: str
    trigger: Callable[[TaskContext], bool]
    message: Callable[[TaskContext], str]


class ReviewGatedAgent:
    def __init__(self):
        self.gates: list[ReviewGate] = []
        self.gate_log: list[dict] = []

    def add_gate(self, gate: ReviewGate) -> "ReviewGatedAgent":
        self.gates.append(gate)
        return self

    async def evaluate_gates(self, ctx: TaskContext, get_approval: Callable) -> bool:
        """TODO: Evaluate all gates, request approval for triggered ones"""
        return True

    async def run_task(self, ctx: TaskContext, task: Callable, get_approval: Callable) -> None:
        """TODO: Evaluate gates then run task if approved"""
        print(f"Running: {ctx.description}")


agent = ReviewGatedAgent()
agent.add_gate(ReviewGate(
    name="many-files",
    trigger=lambda ctx: len(ctx.files_modified) > 5,
    message=lambda ctx: f"⚠️ This task modifies {len(ctx.files_modified)} files.",
))
agent.add_gate(ReviewGate(
    name="db-migration",
    trigger=lambda ctx: ctx.includes_db_migration,
    message=lambda ctx: "⚠️ Database schema change detected.",
))


async def simulate_approval(message: str) -> bool:
    print(f"\n[Review]\n{message}\n[Auto-approve: YES]")
    return True


async def main():
    await agent.run_task(
        TaskContext("Fix typo", ["README.md"], False, False, 1),
        lambda: print("  ✓ Fixed"),
        simulate_approval
    )
    await agent.run_task(
        TaskContext("Refactor auth", [f"auth/f{i}.ts" for i in range(8)], False, False, 340),
        lambda: print("  ✓ Refactored"),
        simulate_approval
    )
    print("\nGate log:", agent.gate_log)


asyncio.run(main())
